android/java/org/chromium/chrome/browser/speedreader/BraveSpeedReaderManager.java

```diff
@@ -0,0 +1,200 @@
+ * message UI.
+ */
+public class BraveSpeedReaderManager extends EmptyTabObserver implements UserData {
+    /** The key to access this object from a {@Tab}. */
+    public static final Class<BraveSpeedReaderManager> USER_DATA_KEY =
+            BraveSpeedReaderManager.class;
+
+    private static final String TAG = "BraveSpeedreaderMgr";
+
+    /** If the infobar was closed due to the close button. */
+    private boolean mIsDismissed;
+
+    /** Whether this manager and tab have been destroyed. */
+    private boolean mIsDestroyed;
+
+    /** The tab this manager is attached to. */
+    private final Tab mTab;
+
+    /** The supplier of MessageDispatcher to display the message. */
+    private final Supplier<MessageDispatcher> mMessageDispatcherSupplier;
+
+    /** Whether the message ui is being shown or has already been shown. */
+    private boolean mMessageShown;
+
+    /** Property Model of Reader mode message. */
+    private PropertyModel mMessageModel;
+
+    BraveSpeedReaderManager(Tab tab, Supplier<MessageDispatcher> messageDispatcherSupplier) {
+        super();
+        mTab = tab;
+        mTab.addObserver(this);
+        mMessageDispatcherSupplier = messageDispatcherSupplier;
+    }
+
+    /**
+     * Create an instance of the {@link BraveSpeedReaderManager} for the provided tab.
+     * @param tab The tab that will have a manager instance attached to it.
+     */
+    public static void createForTab(Tab tab) {
+        tab.getUserDataHost().setUserData(USER_DATA_KEY,
+                new BraveSpeedReaderManager(
+                        tab, () -> MessageDispatcherProvider.from(tab.getWindowAndroid())));
+    }
+
+    /**
+     * Clear the status map and references to other objects.
+     */
+    @Override
+    public void destroy() {
+        mIsDestroyed = true;
+    }
+
+    @Override
+    public void onShown(Tab shownTab, @TabSelectionType int type) {
+        // If the reader infobar was dismissed, stop here.
+        if (mIsDismissed) return;
+
+        tryShowingPrompt();
+    }
+
+    @Override
+    public void onDestroyed(Tab tab) {
+        if (tab == null) return;
+
+        removeTabState();
+    }
+
+    @Override
+    public void onActivityAttachmentChanged(Tab tab, @Nullable WindowAndroid window) {
+        // Intentionally do nothing to prevent automatic observer removal on detachment.
+    }
+
+    /** Clear the reader mode state for this manager. */
+    private void removeTabState() {
+        mIsDismissed = false;
+    }
+
+    /** Try showing the reader mode prompt. */
+    void tryShowingPrompt() {
+        if (!isEnabled()) return;
+        if (mTab == null || mTab.getWebContents() == null) return;
+
+        // Test if the user is requesting the desktop site.
+        boolean usingRequestDesktopSite =
+                mTab.getWebContents().getNavigationController().getUseDesktopUserAgent();
+
+        if (usingRequestDesktopSite || mIsDismissed
```

Comment:
why can't we use it while in a desktop view mode?

Reply:
let's remove it and check what happens? Maybe there are complications with an upstream simplified view in that view mode? But we should be ok with our Speedreader implementation.